### PR TITLE
fix(forge) Prefund test/script contracts before deployment

### LINKED ISSUE #6293: Unable to transfer eth from test constructor
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [ ] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (34f684d 2023-11-01T22:19:24.989266000Z)

### What command(s) is the bug in?

forge test

### Operating System

None

### Describe the bug

The following code reverts.
(on a very strange error, `[FAIL. Reason: EvmError: Revert] setUp() (gas: 0)`, even though there is no "setUp()")

Changing the "constructor" to "setUp" works.

```solidity
contract TestBalance is Test {

    address addr = address(0x1234);

    function initBalance() internal {
        payable(addr).transfer(1 ether);
    }
//    function setUp() public {
    constructor() {
        initBalance();
    }
}
```

### PR DESCRIPTION
**Pull request by klkvr**
## Motivation

Closes #6293 

## Solution

Precompute the test/script contract address and fund it with ETH before deployment

### CODE REVIEW

File: crates/forge/src/runner.rs [hunk 1]
```diff
@@ -113,22 +113,26 @@ impl<'a> ContractRunner<'a> {
...
-                    return Ok(TestSetup::from_evm_error_with(e, logs, traces, Default::default()))
-                }
-            };
+        let sender_nonce = self.executor.get_nonce(self.sender)?;
+        let address = self.sender.create(sender_nonce);
 
-        // Now we set the contracts initial balance, and we also reset `self.sender`s and `CALLER`s
-        // balance to the initial balance we want
+        // Set the contracts initial balance before deployment, so it is available during
+        // construction
```
Comment (mattsse) [inline crates/forge/src/runner.rs:120]: okay this simply switches the order of things.

File: crates/forge/bin/cmd/script/runner.rs [hunk 2]
```diff
@@ -71,20 +71,20 @@ impl ScriptRunner {
             .map(|traces| (TraceKind::Deployment, traces))
             .collect();
 
+        let sender_nonce = self.executor.get_nonce(CALLER)?;
+        let address = CALLER.create(sender_nonce);
+
+        // Set the contracts initial balance before deployment, so it is available during the
+        // construction
+        self.executor.set_balance(address, self.initial_balance)?;
```
Comment (mattsse) [inline crates/forge/bin/cmd/script/runner.rs:79]: this seems reasonable.

File: crates/evm/evm/src/executors/mod.rs [hunk 3]
```diff
@@ -138,6 +138,11 @@ impl Executor {
         Ok(self)
     }
 
+    /// Gets the nonce of an account
+    pub fn get_nonce(&self, address: Address) -> DatabaseResult<u64> {
+        Ok(self.backend.basic(address)?.map(|acc| acc.nonce).unwrap_or_default())
```
Comment (mattsse) [inline crates/evm/evm/src/executors/mod.rs:143]: we need to rebase master, where this will be called `basic_ref` instead

**Review by mattsse — APPROVED** [review]
lgtm thanks.

this sounds very reasonable to me 
wdyt @Evalir?

needs rebase

**Review by Evalir — APPROVED** [review]
Right so, I agree with this for sure. However, I wanna loop @mds1 because maybe some people rely on the test/script contracts not being prefunded to check any operations? We might inadvertently break tests if we merge this. So, to summarize:
- Should we document what the prefund amount is?
- Will this break any existing tests?

**Comment by mds1** [comment]
Is the change here that:
- before we deployed the test (or script) contract, then once deployed it was funded with ETH, then after funding `setup` was executed
- after this PR we fund first, then deploy, then execute `setUp`?

If so I don’t think that should break anything because you can’t rely on a non-funded test/script contract in either case 

However it’s also worth noting that I don’t think we officially support doing things in the constructor for the test contract so relying on it might cause other issues.

**Review by Evalir — APPROVED** [review]

**Pull request merged**